presto-parser/src/main/java/io/prestosql/sql/parser/ErrorHandler.java

```diff
@@ -199,14 +174,59 @@ public Analyzer(
             this.ignoredRules = ignoredRules;
         }
 
-        public Multimap<Integer, String> process(ATNState currentState, int tokenIndex, RuleContext context)
+        public Result process(ATNState currentState, int tokenIndex, RuleContext context)
+        {
+            RuleStartState startState = atn.ruleToStartState[currentState.ruleIndex];
+
+            if (isReachable(currentState, startState)) {
+                // We've been dropped inside a rule in a state that's reachable via epsilon transitions. This is,
+                // effectively, equivalent to starting at the beginning (or immediately outside) the rule.
+                // In that case, backtrack to the beginning to be able to take advantage of logic that remaps
+                // some rules to well-known names for reporting purposes
+                currentState = startState;
+            }
+
+            Set<Integer> endTokens = process(new ParsingState(currentState, tokenIndex, false, parser), 0);
+            Set<Integer> nextTokens = new HashSet<>();
+            while (!endTokens.isEmpty() && context.invokingState != -1) {
+                for (int endToken : endTokens) {
+                    ATNState nextState = ((RuleTransition) atn.states.get(context.invokingState).transition(0)).followState;
+                    nextTokens.addAll(process(new ParsingState(nextState, endToken, false, parser), 0));
+                }
+                context = context.parent;
+                endTokens = nextTokens;
+            }
+
+            return new Result(furthestToken, candidates);
+        }
+
+        private boolean isReachable(ATNState target, RuleStartState from)
         {
-            return process(new ParsingState(currentState, tokenIndex, false, makeCallStack(context), parser, 0));
+            Deque<ATNState> activeStates = new ArrayDeque<>();
+            activeStates.add(from);
+
+            while (!activeStates.isEmpty()) {
+                ATNState current = activeStates.pop();
+
+                if (current.stateNumber == target.stateNumber) {
```

Comment:
The search could be shorter if you checked the condition before _adding_ to the queue, not after _pop_. (just thinking)

Reply:
I'd need to duplicate the check in the `add` above the loop (in case target == from) and before the `push()` below. I thought it unnecessarily complicated the code for very little practical gain.